src/zkevm_specs/evm_circuit/util/precompile_gadget.py

```diff
@@ -0,0 +1,36 @@
+from zkevm_specs.evm_circuit.precompile import Precompile
+from ...util import FQ
+from ..instruction import Instruction
+
+
+class PrecompileGadget:
+    address: FQ
+
+    def __init__(
+        self,
+        instruction: Instruction,
+        callee_addr: FQ,
+        precompile_return_len: FQ,
+        calldata_len: FQ,
+    ):
+        # next execution state must be one of precompiles
+        instruction.constrain_equal(instruction.precompile(callee_addr), FQ.one())
+
+        ### precompiles' specific constraints
+        precompile = Precompile(callee_addr)
+        if precompile == Precompile.DATACOPY:
+            # input length is the same as return data length
+            instruction.constrain_equal(precompile_return_len, calldata_len)
+        elif precompile == Precompile.ECRECOVER:
+            # input length is 128 bytes
+            instruction.constrain_equal(calldata_len, FQ(128))
```

Comment:
It might be problematic to constrain equal calldata len here.
The precopmile runs successfully when the calldata len is 128 bytes.
But EVM allows us to fill in other calldata len and then fail the precompile run.

For example, in this [playground](https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code=%27jFirsNplace_parameters%20in%20memoryZ456e9aea5e197a1f1af7a3e85a3212fa4049a3ba34c2289b4c860fc0b0c64ef3whash~Y~28wvX2YZ9242685bf161793cc25603c231bc2f568eb630ea16aa137d2664ac8038825608wrX4YZ4f8ae3bd7535248d0bd448298cc2e2071e56992d0774dc340c368ae950852adawsX6YqqjDo_call~32JSizeX80JOffsetX8VSize~VOffset~1waddressW4QFFFFFFFFwgasqSTATICCALLqqjPut_resulNalonKon_stackqPOPX80qMLOAD%27~W1%20w%20jq%5Cnj%2F%2F%20_%20thKZW32QY0qMSTOREX~0xWqPUSHV0wargsQ%200xNt%20Ke%20Jwret%01JKNQVWXYZ_jqw~_) I tried to change the argsSize from 0x80 to 0x7f and the precompile fails. That suggests the wrong call data length is still allowed in EVM.

It might take much effort to fix the issue in this PR; we can consider fixing it later.

Reply:
TBH, I'm not pretty sure what constraints we need here. From [Scroll's PR](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1628/files#diff-a5cfbb05b952ef55b7e2fbb8b88a437aa6fd29b9cc5fac042011cb645541e831R54-R63), it only checks  the equality of `precompile_return_len` and `calldata_len`. I doubt that we need any constraints for `ecRecover` or we need to constrain the data consistency between the inputs in each precompile and the data in `callop`. (In Scroll codebase, they're using rlced data and they [check rlced data consistency in precompile_gadget](https://github.com/scroll-tech/zkevm-circuits/blob/57a2dd4c4f8b0443a4cc9ee43c71be649c937ffc/zkevm-circuits/src/evm_circuit/util/precompile_gadget.rs#L58-L80)).

An issue created, https://github.com/privacy-scaling-explorations/zkevm-specs/issues/509

Reply:
fixed in [3cf679c](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/508/commits/3cf679c68644488840d756cd3dadac1f2ce1f293), but I guess the return length you were saying is `32` instead of 128, right.

Reply:
Ah yes!